hail/src/main/scala/is/hail/expr/types/physical/PBetterLocus.scala

```diff
@@ -0,0 +1,90 @@
+package is.hail.expr.types.physical
+
+import is.hail.annotations._
+import is.hail.asm4s._
+import is.hail.expr.ir.{EmitCodeBuilder, EmitMethodBuilder}
+import is.hail.variant._
+
+final case class PBetterLocus(rgBc: BroadcastRG, required: Boolean = false) extends PLocus {
+  def rg: ReferenceGenome = rgBc.value
+
+  def _asIdent = "locusplus"
+
+  // 32 high bits are the contig, 32 low bits is the position
+  val representation: PInt64 = PInt64(required)
+
+  override def _pretty(sb: StringBuilder, indent: Call, compact: Boolean): Unit = sb.append(s"PBetterLocus($rg)")
+
+  private def contigIndex(value: Long): Int = (value >>> 32).toInt
+
+  def contig(value: Long): String = rg.contigs(contigIndex(value))
+
+  def contigType: PString = PCanonicalString(required = true)
+
+  def positionType: PInt32 = PInt32(true)
+
+  def codeOrdering(mb: EmitMethodBuilder[_], other: PType): CodeOrdering = {
+    assert(other isOfType this)
+    new CodeOrderingCompareConsistentWithOthers {
+      type T = Long
+
+      override def compareNonnull(lhs: Code[Long], rhsc: Code[Long]): Code[Int] = {
+        val rhs = other match {
+          case _: PBetterLocus => rhsc
+          case t: PCanonicalLocus => t.toBetterLocus(mb, rhsc)
+        }
+
+        Code.invokeStatic[java.lang.Long, Long, Long, Int]("compare", lhs, rhs)
+      }
+    }
+  }
+
+  def setRequired(required: Boolean): PType = if (this.required == required)
+    this
+  else
+    PBetterLocus(this.rgBc, required)
+
+  private[physical] def codeContigs(mb: EmitMethodBuilder[_]): Code[Array[String]] = {
+    mb.getReferenceGenome(rg).getField[Array[String]]("contigs")
+  }
+}
+
+class PBetterLocusSettable(val pt: PBetterLocus,
+  val value: Settable[Long]
+) extends PLocusValue with PSettable {
+  def get = new PBetterLocusCode(pt, value)
+
+  def store(pc: PCode): Code[Unit] = {
+    value := pc.asInstanceOf[PBetterLocusCode].value
+  }
+
+  def contig(mb: EmitMethodBuilder[_]): PStringCode =
+    new PJavaStringCode(PCanonicalString(true), pt.codeContigs(mb)((value >>> 32).toI))
+
+  def position(): Code[Int] = value.toI
+}
+
+class PBetterLocusCode(val pt: PBetterLocus, val value: Code[Long]) extends PLocusCode {
+  def code: Code[_] = value
+
+  def contig(mb: EmitMethodBuilder[_]): PStringCode =
```

Comment:
if you use a Literal instead, this might be more efficient -- you won't take the string and put it in a region, it will already be there.

Reply:
I don't understand what you're saying. This string is pulled out of the reference genome's contigs array. It's already defined. We never stick it into a region unless the pipeline needs it to be.

Reply:
what are the operations we do with strings in IR? Which of those create java strings, which don't? 

I contend that the common operations (equality, set containment, annotation) all operate on the region representations.

If we do `ht.filter(ht.locus.contig == '22')`, the '22' will be stored as a literal in a region. Using a literal for `contig` here would mean that our PString comparator (length + byte equality) would operate on two values already in regions. Using a java string means that there has to be a converiosn.

Reply:
there will certainly be times where it's more efficient to use a java string like this, but I think it's not the common case.,

Reply:
literal also has the advantage of not needing to serialize a reference genome.

Reply:
Okay. So what you're proposing is to compile, `Literal(TArray(TString), rg.contigs)` into the class/module. And then use that rather than pulling from the reference genome directly.

I'm all for not serializing a reference genome if we don't have to.